ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs

```diff
@@ -175,7 +186,20 @@ initNodeKernel args@NodeKernelArgs { registry, cfg, tracers
       , getFetchMode           = readFetchMode blockFetchInterface
       , getNodeCandidates      = varCandidates
       , getTracers             = tracers
+      , setBlockForging        = \a -> atomically . LazySTM.writeTVar blockForgingVar $! a
       }
+  where
+    blockForgingController :: InternalState m remotePeer localPeer blk
+                           -> STM m [BlockForging m blk]
+                           -> m Void
+    blockForgingController st getBlockForging = go []
+      where
+        go :: [Thread m Void] -> m Void 
+        go !forgingThreads = do
+          blockForging <- atomically getBlockForging
+          traverse_ cancelThread forgingThreads
+          blockForging' <- traverse (forkBlockForging st) blockForging
```

Comment:
I think it's good that we wait until all threads are dead before we spawn any new threads. 👍

Reply:
How this will work if the forging thread is creating a block, would that block be created or would it be discarded?

Reply:
so we need something like:

```haskell
 bracket (atomically $ ChainDB.addBlockAsync ...)
         (\_ -> return ())
         (\result -> do curTip <- atomically $ ChainDB.blockProcessed result
                        ...
                        lift $ removeTxs ...
                        
         )
```
For this to be safe in presence of async exceptions the finaliser must be either a non blocking call, or we need to use `uninterruptibleMask`.  I suspect `ChainDB.blockProcessed` is a blocking call, `removeTxs` if it has access to some `TMVar` could also be blocking.

`addBlockAsync` is quite likely non-blocking (is it?), `bracket` will mask exceptions when it is executed so an async exception will not hit when it's running.

Reply:
This is the last remaining issue -- otherwise the diff looks good to me.

Reply:
This is fixed, although not with a `bracket` but with `uninterruptibleMask_`.

Reply:
See the [commit](https://github.com/input-output-hk/ouroboros-network/pull/3800/commits/7356c6431c79fc1fa4c15273b9761e31086e2339).

Reply:
That logic sounds good to me, with one potential caveat, which I spelled out in this other Thread: https://github.com/input-output-hk/ouroboros-network/pull/3800#discussion_r997642861 But, as I suggested there, I think it's OK as-is. Feels too subtle though :(